crypto/params.c

```diff
@@ -0,0 +1,758 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include "internal/thread_once.h"
+
+#define SET_RETURN_SIZE(p, sz) \
+    if ((p)->return_size != NULL) \
+        *(p)->return_size = (sz)
+
+const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static OSSL_PARAM ossl_param_construct(const char *key, unsigned int data_type,
+                                       void *buffer, size_t buffer_size,
+                                       size_t *return_size)
+{
+    OSSL_PARAM res;
+
+    res.key = key;
+    res.data_type = data_type;
+    res.buffer = buffer;
+    res.buffer_size = buffer_size;
+    res.return_size = return_size;
+    return res;
+}
+
+int OSSL_PARAM_get_int(const OSSL_PARAM *p, int *val)
+{
+    if (val == NULL
+        || p == NULL
+        || (p->data_type != OSSL_PARAM_INTEGER
+            && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER))
+         return 0;
+    switch (p->buffer_size) {
+    case sizeof(int32_t):
```

Comment:
How likely do we think it is that someone will use a BIGNUM for a parameter that is *defined* to be 32-bit integers?

@paulidale's code did originally cover very small sizes, if I recall correctly, but was remarked upon as "a bit too much" or something like that...  perhaps by yours truly, even (the details are blurry).
If you think it's worth it to handle those small sizes, it's not much or hard work to add the few lines that handle them.

Reply:
I do think they should be handled.

Reply:
Ok, I'll add the lines

Reply:
This *might* be over engineering...  how often do you expect that we will see someone use bignums for a documented 32- or 64-bit integer?

Reply:
The more I think about this, the more I think this *does* get us to over engineer it all.
I would rather expect that if an object parameter is defined to be number of some specific set of bits, everyone will have to use that size, or *may* use any power-of-2 size above that.  If the actual number is implemented as a bignum doesn't really matter, it will simply have to be exported in a fitting size.
However, if the number is defined to be multi-precision, then it's fully expected to implemented as an export/import of bignums.

Reply:
I suggest we defer the arbitrary size discussion for later. I don’t think adding it later will be an API breaking change.

Reply:
I'm not convinced that we won't need to do this. But, ok to defer it for now

Reply:
I agree, I expect we will want narrowing and signedness conversions eventually.